classes/ActionScheduler_WPCLI_Scheduler_command.php

```diff
@@ -13,44 +13,116 @@ class ActionScheduler_WPCLI_Scheduler_command extends WP_CLI_Command {
 	 * [--batch-size=<size>]
 	 * : The maximum number of actions to run. Defaults to 100.
 	 *
+	 * [--batches=<size>]
+	 * : Limit execution to a number of batches. Defaults to 0, meaning batches will continue being executed until all actions are complete.
+	 *
 	 * [--force]
 	 * : Whether to force execution despite the maximum number of concurrent processes being exceeded.
+	 *
+	 * @param array $args Positional arguments.
+	 * @param array $assoc_args Keyed arguments.
+	 * @throws \WP_CLI\ExitException When an error occurs.
 	 */
 	public function run( $args, $assoc_args ) {
 		// Handle passed arguments.
-		$batch = \WP_CLI\Utils\get_flag_value( $assoc_args, 'batch-size', 100 );
-		$force = \WP_CLI\Utils\get_flag_value( $assoc_args, 'force', false );
+		$batch   = absint( \WP_CLI\Utils\get_flag_value( $assoc_args, 'batch-size', 100 ) );
+		$batches = absint( \WP_CLI\Utils\get_flag_value( $assoc_args, 'batches', 0 ) );
+		$force   = \WP_CLI\Utils\get_flag_value( $assoc_args, 'force', false );
 
-		$completed = 0;
+		$batches_completed = 0;
+		$actions_completed = 0;
+		$unlimited         = $batches === 0;
 
 		try {
 			// Get the queue runner instance
 			$runner = new ActionScheduler_WPCLI_QueueRunner();
 
-			// Determine how many tasks will be run.
+			// Determine how many tasks will be run in the first batch.
 			$total = $runner->setup( $batch, $force );
-			WP_CLI::line(
-				sprintf(
-					_n( 'Found %d scheduled task', 'Found %d scheduled tasks', $total, 'action-scheduler' ),
-					number_format_i18n( $total )
-				)
-			);
-
-			$completed = $runner->run();
+
+			// Run actions for as long as possible.
+			while ( $total > 0 && ( $unlimited || $batches_completed < $batches ) ) {
+				$this->print_total_actions( $total );
+				$actions_completed += $runner->run();
+				$batches_completed++;
+
+				// Set up tasks for the next batch.
+				$total = $runner->setup( $batch, $force );
```

Comment:
`$runner->setup()` claims actions, which means in the last iteration of the loop, a batch of actions is claimed, but never run, because the `while`'s batch count condition fails and `$runner->run()` is not called.

Initially, I thought the fix might be as simple as replacing `$total` with the `$total = $runner->setup()` assignment in the `while` conditions. That will still claim actions though, because `$runner->setup()` is called even when `$batches_completed < $batches` proves false, so the order of those two conditions will also need to be switched to make sure the batch count condition fails before claiming anymore actions, e.g.

```php
// Run actions for as long as possible.
while ( ( $unlimited || $batches_completed < $batches ) && ( $total = $runner->setup( $batch, $force ) ) > 0 ) {
```

Reply:
@thenbrent Good catch!

An alternate solution might be to add a `__destruct()` method to the `ActionScheduler_WPCLI_QueueRunner` class:

```php
public function __destruct() {
    $this->store->release_claim( $this->claim );
}
```

This would generally ensure that the claim is released regardless of the reason for the process stopping, although of course there might be exceptions for certain kinds of failures.

Reply:
> regardless of the reason for the process stopping

@JPry generally, we want the claim to remain if the process stopped for unknown reasons so we can mark `in-progress` actions as failed after the timeout period. Releasing the claim anytime `ActionScheduler_WPCLI_QueueRunner::__desctruct()` is called could lead some actions to fail silently.

Leaving the bug in `run()` and addressing it outside of there can also create other problems (that will be difficult to diagnose). For example, if the claim changes between when `ActionScheduler_WPCLI_QueueRunner::run()` is called and when `ActionScheduler_WPCLI_QueueRunner::__desctruct()` is called, which could (unintentionally) happen seeing as the claim is staked and set on the instance in `ActionScheduler_WPCLI_QueueRunner::setup()`, which can be called independently of `run()`.

Reply:
Ah-hah, that makes sense. I can see why the `__destruct()` method isn't the best here. I'll post another commit with some updates in the morning.